### PR TITLE
[Org Browser] Hook up metadata to tree view

### PR DESCRIPTION
**Pull request by AnanyaJha**
### What does this PR do?
This PR retrieves the metadata and displays it in the org browser. Adds refresh functionality at the top level. Org browser will automatically refresh with data when the default org has switched. Tests have not been added yet. The org browser is hidden by default; to review, change 'display_tree_view' to 'true'.

### What issues does this PR fix or reference?
@W-5767978@

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 1]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
+      this.defaultOrg = username;
+      return diff;
+    } else {
+      throw new Error('Workspace could not be found.');
     }
   }
 }
+
+export function parseErrors(error: string): string {
+  const e = JSON.parse(error);
```
Comment (AnanyaJha) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Do we want to change the error message based on the type of error?

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts [hunk 2]
```diff
@@ -138,31 +81,27 @@ export type MetadataObject = {
...
-}
-
-export async function onUsernameChange() {
-  const metadataTypesPath = await getTypesPath();
-  if (
-    !isNullOrUndefined(metadataTypesPath) &&
-    !fs.existsSync(metadataTypesPath)
-  ) {
-    await forceDescribeMetadata(metadataTypesPath);
+  const jsonObject = JSON.parse(metadataFile);
```
Comment (AnanyaJha) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: Wasn't sure if I should wrap the JSON parsing here and for the components in a try/catch

File: packages/salesforcedx-vscode-core/package.json [hunk 3]
```diff
@@ -625,6 +632,14 @@
       {
         "command": "sfdx.internal.lightning.lwc.create",
         "title": "%force_lightning_lwc_create_text%"
+      },
+      {
+        "command": "sfdx.force.metadata.view.refresh",
+        "title": "Refresh tests",
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/package.json]: This needs to be a label and the text should be something along the lines of `Refresh Org Browser`. cc @hbc-bindu

File: packages/salesforcedx-vscode-core/src/index.ts [hunk 4]
```diff
@@ -490,8 +496,8 @@ export async function activate(context: vscode.ExtensionContext) {
   const orgList = new OrgList();
   orgList.displayDefaultUsername(defaultUsernameorAlias);
   context.subscriptions.push(registerOrgPickerCommands(orgList));
-  // await setupOrgBrowser(context, defaultUsernameorAlias);
 
+  await setupOrgBrowser(context, defaultUsernameorAlias);
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/index.ts]: I think we should keep this commented out since we're always passing `false` for the tree view context in the following line.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 5]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
+    }
+    const nodeList = [];
+    for (const component of componentsList) {
+      const cmpNode = new BrowserNode(component, NodeType.MetadataCmp);
+      nodeList.push(cmpNode);
+    }
+    return nodeList;
+  }
+
+  public async getDefaultUsernameOrAlias(): Promise<boolean> {
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: The name and responsibility of this method is a bit confusing. We use it to determine if the user has switched to a different org and also to set the username for the rest of the org browser code to leverage. Can we break this into different methods and name them based on what they are responsible for ?

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 6]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
-  public async refresh(defaultOrg?: string): Promise<void> {
-    if (!isNullOrUndefined(defaultOrg)) {
-      this.defaultOrg = defaultOrg;
+  public async onViewChange() {
+    if (await this.getDefaultUsernameOrAlias()) {
+      this.internalOnDidChangeTreeData.fire();
     }
+  }
+
+  public async refresh(defaultOrg?: string): Promise<void> {
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Can we visually indicate to the user that a refresh of the org browser data is happening ? Right now, if a user clicks the refresh button there's nothing that lets the user know if the feature is actually refreshing or when it's done refreshing the data.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts [hunk 7]
```diff
@@ -138,31 +81,27 @@ export type MetadataObject = {
...
-export function buildTypesList(metadataTypesPath: string): string[] {
-  try {
-    const fileData = JSON.parse(fs.readFileSync(metadataTypesPath, 'utf8'));
-    const metadataObjects = fileData.metadataObjects as MetadataObject[];
-    const metadataTypes = [];
-    for (const metadataObject of metadataObjects) {
-      if (!isNullOrUndefined(metadataObject.xmlName)) {
-        metadataTypes.push(metadataObject.xmlName);
-      }
+export function buildTypesList(
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: This method should return a string array in alphabetical order. Currently the types are shown in an order that's difficult to navigate.
Screen Shot 2019-06-24 at 7 12 05 PM

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts (hunk 2 above)
Comment (lcampos) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: I think we should expand the try/catch to cover the rest of the code since it's also navigating a json structure that might throw a null/empty value.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts [hunk 8]
```diff
@@ -138,31 +81,27 @@ export type MetadataObject = {
...
+  metadataFile?: any,
+  metadataTypesPath?: string
+): string[] {
+  if (isNullOrUndefined(metadataFile)) {
+    try {
+      metadataFile = fs.readFileSync(metadataTypesPath!, 'utf8');
+    } catch (e) {
+      throw e;
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: Need to log the exception as well.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts [hunk 9]
```diff
@@ -8,124 +8,67 @@
 import {
   CliCommandExecutor,
   Command,
+  CommandOutput,
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: The code in this file both runs a describe command and hosts utility methods (e.g. getTypesFolder) that are unrelated to the describe command. Can we create another file specific for the utility methods used across the org browser code ?

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 1 above)
Comment (lcampos) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Yes, what errors are not already covered here ?

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 10]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
     }
   }
 }
+
+export function parseErrors(error: string): string {
+  const e = JSON.parse(error);
+  let message: string;
+  switch (e.name) {
+    case 'RefreshTokenAuthError':
+      message = 'Error refreshing authentication token.';
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Let's update this messages to use labels. cc @hbc-bindu

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 11]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
+    await this.getDefaultUsernameOrAlias();
     this.internalOnDidChangeTreeData.fire();
   }
 
   public getTreeItem(element: BrowserNode): vscode.TreeItem {
     return element;
   }
 
-  public getChildren(element?: BrowserNode): Promise<BrowserNode[]> {
+  public async getChildren(element?: BrowserNode): Promise<BrowserNode[]> {
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Let's rework this method to make it easier to read.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 1 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: This was all of them! Was just checking if we wanted these.

File: packages/salesforcedx-vscode-core/src/index.ts (hunk 4 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/index.ts]: Ah yes, was meant to be commented still. Thanks

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 10 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Sounds good, let me know if these messages look okay @hbc-bindu and I'll add them to the i8n file

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts (hunk 9 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: Yeah I can do that- I'll do the same for the metadataCmp file because it also contains utility code and the list metadata command.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 6 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Hmm yeah that makes sense. I'll look into how to indicate the progress.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataCmp.ts [hunk 12]
```diff
@@ -51,115 +36,88 @@ export class ForceListMetadataExecutor extends SfdxCommandletExecutor<string> {
...
-    const fileData = fs.readFileSync(componentsPath, 'utf8');
-    if (fileData !== 'undefined') {
-      const fileObject = JSON.parse(fileData);
-      const cmpList = fileObject instanceof Array ? fileObject : [fileObject];
-      for (const component of cmpList) {
-        if (!isNullOrUndefined(component.fullName)) {
-          metaComponents.push(component.fullName);
-        }
+  const components = [];
+  if (isNullOrUndefined(componentsFile)) {
```
Comment (brpowell) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataCmp.ts]: `util.isNullOrUndefined` has been deprecated for awhile, we should avoid using it going forward.
https://nodejs.org/api/util.html#util_util_isnullorundefined_object

Comment (lcampos) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataCmp.ts]: Good catch

File: packages/salesforcedx-vscode-core/package.json (hunk 3 above)
Comment (hbc-bindu) [reply packages/salesforcedx-vscode-core/package.json]: Label looks good to me.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 10 above)
Comment (hbc-bindu) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Messages are good; no changes.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 13]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
+      nodeList.push(typeNode);
+    }
+    return nodeList;
+  }
+
+  public async getComponents(
+    metadataType: BrowserNode
+  ): Promise<BrowserNode[]> {
+    const username = this.defaultOrg!;
+    const componentsPath = await getComponentsPath(
```
Comment (brpowell) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: I'm wondering if metadataCmp.ts can be responsible for doing more of this work. The provider is really only concerned with having a list of the component names to construct the nodes and shouldn't care whether it comes from the cache or remote.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 6 above)
Comment (brpowell) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Whether there's an org switch or refresh, you could clear the tree and present a non-collapsable "loading" node.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts (hunk 7 above)
Comment (brpowell) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: Icons?
Screen Shot 2019-06-25 at 2 51 41 PM

No Icons?
Screen Shot 2019-06-25 at 2 52 30 PM

@ntotten

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 14]
```diff
@@ -24,33 +37,144 @@ export class MetadataOutlineProvider
...
+      }
+    } else if (element.type === NodeType.Org) {
+      const metadataTypes = await this.getTypes();
+      element.children = metadataTypes;
+      return Promise.resolve(element.children);
+    } else if (element.type === NodeType.MetadataType) {
+      const metadataCmps = await this.getComponents(element);
+      if (isNullOrUndefined(metadataCmps) || metadataCmps.length === 0) {
+        const emptyDefault = new BrowserNode(
+          'No components available',
```
Comment (brpowell) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Do we want to show a node if there aren't any components? In the normal file tree if there aren't any files in a folder nothing is shown.

**Comment by AnanyaJha** [comment]
Still working on visually indicating a refresh is in progress and the `display tree view` context is set to true for now.

File: packages/salesforcedx-utils-vscode/src/helpers/utils.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2019, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+export function isEmpty(object: any): object is null | undefined {
```
Comment (lcampos) [inline packages/salesforcedx-utils-vscode/src/helpers/utils.ts]: 👍

File: packages/salesforcedx-vscode-core/src/commands/forceListMetadata.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    this.metadataType = metadataType;
+    this.defaultUsernameOrAlias = defaultUsernameOrAlias;
+  }
+
+  public build(data: {}): Command {
+    let builder = new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_list_metadata_text'))
+      .withArg('force:mdapi:listmetadata')
+      .withFlag('-m', this.metadataType)
+      .withFlag('-u', this.defaultUsernameOrAlias)
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceListMetadata.ts]: Let's add a `.withLogName` entry to this command.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 14 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: I'm leaning towards displaying a node with a `No components available` label because I found it difficult to easily see that the type node was expanded and no components were below it once the data had been requested.

cc @ntotten Thoughts?

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 17]
```diff
@@ -20,37 +21,107 @@ export class MetadataOutlineProvider
...
+        break;
+      case NodeType.MetadataType:
+        element.children = await this.getComponents(element);
+        break;
+    }
+    return Promise.resolve(element.children);
   }
 
-  public async getDefaultUsernameOrAlias() {
+  public async getTypes(): Promise<BrowserNode[]> {
```
Comment (brpowell) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: getTypes(), getComponents(), and getDefaultUsernameOrAlias() don't need to be public.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts [hunk 18]
```diff
@@ -20,37 +21,107 @@ export class MetadataOutlineProvider
...
-      this.defaultOrg = await OrgAuthInfo.getDefaultUsernameOrAlias(false);
+      const username = await OrgAuthInfo.getDefaultUsernameOrAlias(false);
+      return username;
+    } else {
+      throw new Error(nls.localize('cannot_determine_workspace'));
     }
   }
 }
+
+export function parseErrors(error: string): string {
```
Comment (brpowell) [inline packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: Does this need to be exported?

File: packages/salesforcedx-utils-vscode/src/helpers/utils.ts (hunk 15 above)
Comment (brpowell) [inline packages/salesforcedx-utils-vscode/src/helpers/utils.ts]: This name is a misnomer. We're not testing if an object is empty, but whether it is null or undefined. An empty object would be `{}`, and this function would return false for such a value. null and undefined are falsy values so it's fine to just test this using bang.
```
> !undefined
true
> !null
true
> !{}
false
```

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 14 above)
Comment (ntotten) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: I don't have a strong opinion on this. The file tree in VS Code allows you to click expand on the node and just doesn't show anything when the folder is empty. In this case though since we are loading data from the org I almost wonder if not showing anything will make it feel like an err. I'm okay with the current design.

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts (hunk 7 above)
Comment (ntotten) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: I personally like the icons. Once we expand to "source format" i think it would be good to use the same icons as the file explorer for xml, js, html, cls, etc.

Comment (ntotten) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: One thought, in the file explorer there is an option to show icons or not. Maybe we can use that same setting?

image

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 14 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: I'll leave the message in for now and we can adjust if we get feedback on it. Thanks!

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts (hunk 17 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts]: A few of the methods need to be public so I can access them for testing.

File: packages/salesforcedx-utils-vscode/src/helpers/utils.ts (hunk 15 above)
Comment (AnanyaJha) [reply packages/salesforcedx-utils-vscode/src/helpers/utils.ts]: I think it's useful to keep this function since we have been using 'isNullOrUndefined' across the project. I'll adjust the name to be more representative of what the function is doing

File: packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts (hunk 7 above)
Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/orgBrowser/metadataType.ts]: The icons look good to me too, I think we can add them in a future story. I'll log a work item for the icons with the option of exploring that setting ^^

File: packages/salesforcedx-vscode-core/src/commands/forceListMetadata.ts [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    this.metadataType = metadataType;
+    this.defaultUsernameOrAlias = defaultUsernameOrAlias;
+  }
+
+  public build(data: {}): Command {
+    let builder = new SfdxCommandBuilder()
+      .withArg('force:mdapi:listmetadata')
+      .withFlag('-m', this.metadataType)
+      .withFlag('-u', this.defaultUsernameOrAlias)
+      .withLogName(nls.localize('force_list_metadata_text'))
```
Comment (lcampos) [inline packages/salesforcedx-vscode-core/src/commands/forceListMetadata.ts]: This label will break the current metric name format we use, let's replace it for using something like `force_list_metadata`.

Comment (AnanyaJha) [reply packages/salesforcedx-vscode-core/src/commands/forceListMetadata.ts]: Sounds good, will do!
